utils/zerovec/src/flexzerovec/slice.rs

```diff
@@ -0,0 +1,264 @@
+        let mut bytes = [0; USIZE_WIDTH];
+        core::ptr::copy_nonoverlapping(self.data.as_ptr().add(index * w), bytes.as_mut_ptr(), w);
+        usize::from_le_bytes(bytes)
+    }
+
+    #[inline]
+    pub fn first(&self) -> Option<usize> {
+        let w = self.get_width();
+        self.data.get(0..w).map(|chunk| chunk_to_usize(chunk, w))
+    }
+
+    #[inline]
+    pub fn last(&self) -> Option<usize> {
+        let l = self.data.len();
+        if l == 0 {
+            None
+        } else {
+            let w = self.get_width();
+            self.data
+                .get(l - w..l)
+                .map(|chunk| chunk_to_usize(chunk, w))
+        }
+    }
+
+    #[inline]
+    pub fn iter(&self) -> impl DoubleEndedIterator<Item = usize> + '_ {
+        let w = self.get_width();
+        self.data
+            .chunks_exact(w)
+            .map(move |chunk| chunk_to_usize(chunk, w))
+    }
+
+    /// Binary searches a sorted `FlexZeroSlice` for the given `usize` value.
+    pub fn binary_search(&self, needle: usize) -> Result<usize, usize> {
+        // See comments in components.rs regarding the following code.
+
+        let zero_index = self.data.as_ptr() as *const _ as usize;
+        debug_assert!(self.len() <= self.data.len());
+        // Safety: self.len() <= self.data.len()
+        let scaled_slice = unsafe { self.data.get_unchecked(0..self.len()) };
+
+        scaled_slice.binary_search_by(|probe: &_| {
+            // Note: `scaled_slice` is a slice of u8
+            let index = probe as *const _ as usize - zero_index;
+            // Safety: we know this is in bounds
+            let actual_probe = unsafe { self.get_unchecked(index) };
+            <usize as Ord>::cmp(&actual_probe, &needle)
+        })
+    }
+}
+
+#[inline]
+fn get_item_width(item_bytes: &[u8; USIZE_WIDTH]) -> usize {
+    USIZE_WIDTH - item_bytes.iter().rev().take_while(|b| **b == 0).count()
+}
+
+pub(crate) struct InsertInfo {
+    pub item_bytes: [u8; USIZE_WIDTH],
+    pub new_width: usize,
+    pub new_count: usize,
+    pub new_data_len: usize,
+}
+
+impl FlexZeroSlice {
+    pub(crate) fn get_insert_info(&self, new_item: usize) -> InsertInfo {
+        let item_bytes = new_item.to_le_bytes();
+        let item_width = get_item_width(&item_bytes);
+        let old_width = self.get_width();
+        let new_width = core::cmp::max(old_width, item_width);
+        let new_count = 1 + (self.data.len() / old_width);
+        let new_data_len = new_count * new_width;
+        InsertInfo {
+            item_bytes,
+            new_width,
+            new_count,
+            new_data_len,
+        }
+    }
+
+    /// This function should be called on a slice with a data array `new_data_len` long
+    /// which previously held `new_count - 1` elements.
+    ///
+    /// After calling this function, all bytes in the slice will have been written.
+    pub(crate) fn insert_impl(&mut self, insert_info: InsertInfo, insert_index: usize) {
```

Comment:
issue: both this and remove() should optimize for the common case where the width isn't changing

Reply:
Hmm. I am willing to do the following:

- On insertion, after inserting the new element, skip copying the rest of the elements if the width isn't changing (note: this skips the whole array for a `push`)
- On removal, skip directly to the index that is being removed if the width isn't changing (note: this skips the whole array for a `pop`)

If possible, I would prefer to avoid having a whole separate code path, which makes the code harder to test and reason about. As written, `shift` and `unshift` (modifying the beginning of the array) require O(N) memcpy operations. We could improve this further by making that be a single memcpy operation, which will definitely have some improvement, but it's still asymptotically an O(N) operation.

Thoughts?

Reply:
Sounds good.

Reply:
Done as described.